Implement `Memory1` (`RULE-8-7-1`)

jeongsoolee09 opened the pull request:
## Description

Implement Memory1 (`RULE-8-7-1`).

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - `RULE-8-7-1`
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed


_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

jeongsoolee09 committed:
Number `Memory` packages

jeongsoolee09 committed:
Add rule description files

jeongsoolee09 committed:
Add Memory1 package files

jeongsoolee09 committed:
Expose malloc, calloc and realloc

jeongsoolee09 committed:
Minor comments

jeongsoolee09 committed:
Checkpoint

jeongsoolee09 committed:
Split out source and sinks into their cases

jeongsoolee09 committed:
Checkpoint

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Checkpoint: Add `DynamicAllocation` case

jeongsoolee09 committed:
First working draft

jeongsoolee09 committed:
Refine into path-problem

jeongsoolee09 committed:
Change `TaintTracking` to `DataFlow`

jeongsoolee09 committed:
Finalize first working draft for stack / heap arrays

jeongsoolee09 committed:
Document code copy and clean up imports

jeongsoolee09 committed:
Add multidimensional arrays alloc'ed on stack

jeongsoolee09 committed:
Add test.cpp and expected test results

jeongsoolee09 committed:
Add exclusion for Memory1.qll

jeongsoolee09 committed:
Adjust precision of existing rule and add a supplementary rule

jeongsoolee09 committed:
Add supplementary query files

jeongsoolee09 committed:
Fix @kind from problem to path-problem

jeongsoolee09 committed:
Copy OutOfBounds.qll to cpp/common/src/codingstandards/cpp/

jeongsoolee09 committed:
Remove unused import codingstandards.cpp.Variable in OutOfBounds.qll

jeongsoolee09 committed:
Add PointerArgumentToCstringFunctionIsInvalid.ql and create testref for it

jeongsoolee09 committed:
Copy test.c from ARR38-C and add strncpy

jeongsoolee09 committed:
Add headers, Add defininitions to headers, remove cases without null terminator, remove file pointer cases

1. Add headers, Adding missing headers: For obvious reasons.
2. Remove cases without null terminator: Both clang and g++ do not permit
   strings to be allocated that are declared to be shorter than the actual
   initializing expression. Since this is a C++ rule, we rule them out.
3. File pointer manipulation functions (e.g. fgets): Not required by the rule.

jeongsoolee09 committed:
Remove testref and add qlref and expected

jeongsoolee09 committed:
Address case of `strncat`

jeongsoolee09 committed:
Remove unused predicate and update .expected

jeongsoolee09 committed:
Fix formatting error on the name property of the package

jeongsoolee09 committed:
Fix formatting of test.cpp

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Fix name formatting

jeongsoolee09 committed:
Merge branch 'jeongsoolee09/MISRA-C++-2023-Memory' of github.com:github/codeql-coding-standards into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Commit these later

jeongsoolee09 committed:
Remove alerts for negative offsets and fix @name property

Too many of the negative offsets alerts were false positives.
We leave it for future work.

jeongsoolee09 committed:
Fix precision in description file

jeongsoolee09 committed:
Fix test case formatting

jeongsoolee09 committed:
Add QLDocs and comments

jeongsoolee09 committed:
Update the locations in an expected file

MichaelRFairhurst commented on cpp/common/src/codingstandards/cpp/OutOfBounds.qll:
```diff
@@ -0,0 +1,1358 @@
+/**
+ * This module provides classes and predicates for analyzing the size of buffers
+ * or objects from their base or a byte-offset, and identifying the potential for
+ * expressions accessing those buffers to overflow.
```
Is this a direct copy? We should probably state that, and/or, list modifications that have been made.

MichaelRFairhurst replied:
In this case, can we have `c/common/src/codingatandards/c/OutOfBounds.qll` import `cpp.OutOfBounds` ? A simple wrapper import file would be a reasonable easy refactor.

(or find each c query that imports OutOfBounds and update those)?

I think it's probably reasonable

jeongsoolee09 replied:
I think a more fundamental solution is to have a radical refactor (that is of enormous magnitude and has to be surgically precise) that makes `cpp` depend on `c`, as C++ is (almost, but 100% in practice) a superset of C.

Then we won't have to keep a copy here which is awkward.

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArgumentToCstringFunctionIsInvalid.ql:
```diff
@@ -0,0 +1,26 @@
+  OOB::BufferAccessLibraryFunctionCall fc, string message, Expr bufferArg, string bufferArgStr,
+  Expr sizeOrOtherBufferArg, string otherStr
+where
+  not isExcluded(fc, OutOfBoundsPackage::libraryFunctionArgumentOutOfBoundsQuery()) and
```
I believe this should be `not isExcluded(Memory1Package::theQueryName())`

jeongsoolee09 replied:
Fixed in [8435f3c](https://github.com/github/codeql-coding-standards/pull/967/commits/8435f3cc1f1b65a7257e024b0d038841b42d637b)!

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,301 @@
+    )
+    or
+    exists(PointerSubExpr pointerSubtraction | pointerSubtraction = this.asPointerArithmetic() |
+      result = -pointerSubtraction.getAnOperand().getValue().toInt()
```
Let's handle this todo, but not how copilot is suggesting, so that we can handle both `p + n` and `n + p`.

MichaelRFairhurst reviewed (changes requested):
Very nice work, Jeongsoo! The work you've put into this really shows, its elegant and focused. In terms of the taint tracking edge cases, I think we can handle those cleanly on your foundation here if we focus on the array-to-pointer conversion cases, which we can talk more about later! Also, I have to say the tests you've made are awesomely comprehensive, nicely done. That is huge and really shows all your attention to detail!

jeongsoolee09 committed:
Change the query in that gets excluded to the right one

jeongsoolee09 committed:
Add complete docstring to `NarrowedHeadAllocationFunctionCall`

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Add experimental query

jeongsoolee09 committed:
Centralize TrackArray::flow(Path) to from-where-select

jeongsoolee09 committed:
Move flowPath to srcSinkLengthMap and add stack_only.cpp

jeongsoolee09 committed:
Bind the two PathNodes in `from` to the pointers

jeongsoolee09 committed:
Add `pointer_only.cpp`

jeongsoolee09 committed:
Add more debugging messages

MichaelRFairhurst committed:
Updates to dataflow nodes & tests

jeongsoolee09 committed:
Add cases and relabel compliant / non-compliant

jeongsoolee09 committed:
Merge branch 'jeongsoolee09/MISRA-C++-2023-Memory' into jeongsoolee09/MISRA-C++-2023-Memory-Experimental

jeongsoolee09 committed:
Update pointer_only.cpp according to test.cpp

jeongsoolee09 committed:
Fix redeclaration issue

jeongsoolee09 committed:
Merge remote-tracking branch 'origin/michaelrfairhurst/update-dataflow-nodes-MISRA-C++-2023-Memory-Experimental' into jeongsoolee09/MISRA-C++-2023-Memory-Experimental

jeongsoolee09 committed:
Add multidimensional_only.cpp

jeongsoolee09 committed:
Add docstrings

jeongsoolee09 committed:
Support the multidimensional array cases

jeongsoolee09 committed:
Finish experimentation with srcSinkLengthMap and merge

jeongsoolee09 commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,301 @@
+  import semmle.code.cpp.ir.IR
+  import semmle.code.cpp.ir.dataflow.internal.SsaInternals as Ssa
+
+  predicate operandToInstructionTaintStep(Operand opFrom, Instruction instrTo) {
```
Good thing we've come up with a better solution than this!

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Update RuleMetadata.qll

jeongsoolee09 committed:
Remove unnecessary files and fix formatting

jeongsoolee09 committed:
Fix formatting and mention that OutOfBounds is a copy

jeongsoolee09 committed:
Add test cases

These contain false positives due to the
limitation of the status quo of the query.

jeongsoolee09 committed:
Format message and finish final draft

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,410 @@
+ * A declaration of a variable that is of an array type.
+ */
+class ArrayDeclaration extends VariableDeclarationEntry {
+  int length;
```
We can probably delete this field

jeongsoolee09 replied:
It's used in `ArrayAllocation::getLength/0`, so I'll dismiss it (I know it's an old comment!).

MichaelRFairhurst reviewed (changes requested):
So close to ready!

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Fix minor formatting in the test file

jeongsoolee09 committed:
Add note to IndirectUninitializedNode

jeongsoolee09 committed:
Add an accumulating logic

jeongsoolee09 committed:
Add some documentation

jeongsoolee09 commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,410 @@
+    sink.getNode() = end.getBasePointerNode()
+  |
+    srcOffset = start.getOffset() and
+    sinkOffset = end.getOffset() and
```
Implemented in [394b7ad](https://github.com/github/codeql-coding-standards/pull/967/commits/394b7ada0c36168d8ba8c71f0e7e0e3c24dc23c7) and documented in [2e4ace6](https://github.com/github/codeql-coding-standards/pull/967/commits/2e4ace6b834dbb0b20bfb7662f1365be0e662ec0).

MichaelRFairhurst replied:
Nice!

Only thought now, do still need `srcOffset`, `sinkOffset` to be in the table / to be predicate parameters?

In the base case, the `srcOffset` and `sinkOffset` come from `start` and `end`, not from `srcSinkLengthMap`.

In the recursive case, the `srcOffset` from the previous iteration is unused `(srcSinkLengthMap(_, start, /*here -> */ _, ...)`. The `sinkOffset` from the previous iteration is only bound to be the new `srcOffset`, which we just determined wasn't used in the next iteration.

`srcOffset` and `sinkOffset` are then also not used by the select

jeongsoolee09 replied:
Yes, that's obviously cleaner. 🙂 But I'd like to leave it this way to inspect all "local variables" from simply quick-evaluating the predicate.

jeongsoolee09 committed:
Add more docs

jeongsoolee09 committed:
Fix formatting error

jeongsoolee09 committed:
Add support for address of arbitrary lvalue exprs

jeongsoolee09 commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,410 @@
+
+newtype TArrayAllocation =
+  TStackAllocation(ArrayDeclaration arrayDecl) or
+  TDynamicAllocation(NarrowedHeapAllocationFunctionCall narrowedAlloc)
```
Implemented to some degree in [44ef266](https://github.com/github/codeql-coding-standards/pull/967/commits/44ef2668d99f96412412efc5a7c50302ffb197d9).

jeongsoolee09 commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,410 @@
+   */
+  int getOffset() {
+    if this.asPointerArithmetic() instanceof PointerSubExpr
+    then result = -this.getOffsetExpr().getValue().toInt()
```
Good point. Introducing range analysis should be careful, otherwise it might generate a lot of noise. This is out of scope of this PR and should be reserved for later.

jeongsoolee09 committed:
Fix formatting

jeongsoolee09 committed:
Update expected results of both queries

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 commented:
Two things to note about the multidimensional arrays:
1. If a row access and the row element access are apart from each other with a function boundary in between, the query loses indirection information. We probably need to add a `level` column to simulate push-pop behavior: taking an indirection edge during initialization pushes and accesses pops them.
2. Currently, the relationship between `FatPointer` and `DataFlow::Node` is one-to-many, if we don't-care the level column. The multidimensional array accesses have duplicate alerts and this is probably what is causing it.

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,451 @@
+ * @precision medium
+ * @problem.severity error
+ * @tags external/misra/id/rule-8-7-1
+ *       scope/system
```
looks like we're missing correctness

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,451 @@
+  /**
+   * Gets the declared length of this array.
+   */
+  int getLength() { result = length }
```
This and the `int length;` on line 26 can be deleted now, right?

jeongsoolee09 replied:
No, but we can have `getLength` compute the `length` from the charpred, thereby removing the class field `int length` and `getLength/0`. That is done in `e8cb6b5`.

MichaelRFairhurst commented on cpp/misra/test/rules/RULE-8-7-1/test.cpp:
```diff
@@ -0,0 +1,555 @@
+      array +
+      4; // NON_COMPLIANT: pointer points more than one beyond the last element
+  int *invalid2 =
+      array - 1; // NON_COMPLIANT: pointer is outside boundary [FALSE_NEGATIVE]
```
No longer a false negative! 🎉

jeongsoolee09 replied:
Removed the label in `ebd1e58`!

MichaelRFairhurst commented on cpp/misra/test/rules/RULE-8-7-1/test.cpp:
```diff
@@ -0,0 +1,555 @@
+  int invalid1 = array[4]; // NON_COMPLIANT: pointer points more than one beyond
+                           // the last element
+  int invalid2 =
+      array[-1]; // NON_COMPLIANT: pointer is outside boundary [FALSE_NEGATIVE]
```
(same)

jeongsoolee09 replied:
Removed the label in `ebd1e58` 🙂

MichaelRFairhurst commented on cpp/misra/test/rules/RULE-8-7-1/test.cpp:
```diff
@@ -0,0 +1,555 @@
+    strcat(buf1, " ");     // NON_COMPLIANT - not null terminated
+    strcat(buf2, " ");     // COMPLIANT
+    strcat(buf3, " ");     // COMPLIANT
+    strcat(buf4, "12345"); // NON_COMPLIANT
```
Hmm, this is actually a FN

jeongsoolee09 replied:
Flagged as an FN at [a21c8ba](https://github.com/github/codeql-coding-standards/pull/967/commits/a21c8ba120d7a24e5fb4c7b7039decd04570ebeb). As I wrote in the commit message, the C queries that depend on `OutOfBounds.qll` consistently miss this case:

- c/misra/src/rules/RULE-21-17/StringFunctionPointerArgumentOutOfBounds.ql
- c/misra/src/rules/RULE-21-18/StringLibrarySizeArgumentOutOfBounds.ql
- c/cert/src/rules/ARR30-C/DoNotFormOutOfBoundsPointersOrArraySubscripts.ql
- c/cert/src/rules/ARR38-C/LibraryFunctionArgumentOutOfBounds.ql

Something to get back to later.

MichaelRFairhurst commented on rule_packages/cpp/Memory1.json:
```diff
@@ -0,0 +1,35 @@
+          "description": "Pointers obtained as result of performing arithmetic should point to an initialized object, or an element right next to the last element of an array.",
+          "kind": "path-problem",
+          "name": "Pointer arithmetic shall not form an invalid pointer",
+          "precision": "medium",
```
I'd be very tempted to say "high" precision.

Nicely done :)

MichaelRFairhurst commented on rule_packages/cpp/Memory1.json:
```diff
@@ -0,0 +1,35 @@
+          "severity": "error",
+          "short_name": "PointerArithmeticFormsAnInvalidPointer",
+          "tags": [
+            "scope/system"
```
tags should probably include correctness and security, for this and below

jeongsoolee09 replied:
Added in [60ddfaa](https://github.com/github/codeql-coding-standards/pull/967/commits/60ddfaafc88a440410494a5b4fed2ac68ba24f6f).

MichaelRFairhurst commented on rule_packages/cpp/Memory1.json:
```diff
@@ -0,0 +1,35 @@
+          "tags": [
+            "scope/system"
+          ]
+        },
```
Consider adding an implementation scope that we only handle constant offsets for increased precision

jeongsoolee09 replied:
Added in [60ddfaa](https://github.com/github/codeql-coding-standards/pull/967/commits/60ddfaafc88a440410494a5b4fed2ac68ba24f6f).

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -0,0 +1,451 @@
+  )
+select end, src, sink,
+  "This pointer accesses element at index " + totalOffset +
+    " while the underlying object has length " + length + "."
```
One small thought on verbiage.

While "Object" is the right term here, generally speaking, I'm not sure devs think of objects as something that has a "length"... I'd say objects have a "size," and that size is in bytes, making "length" perhaps doubly confusing.

If you want to go with "length," I'd probably suggest "array" or "buffer." Though it could be confusing to say that `&x` is either.

Maybe something like, "Pointer formed that points to element X of an object contains Y elements" ?

jeongsoolee09 replied:
It's pretty tricky because "length" denotes the count of the elements while "size" is measured in bytes. The more relevant notion to this rule is length that goes together with "indices".

But again, "elements" belongs to containers and not simply any objects. (the thing that contains Y elements is a _container_!)

All in all, I don't think there's a sensible alternative that can satisfy everything. The rule itself is conducive to conflating object with an array.

MichaelRFairhurst reviewed (changes requested):
literally just minor tweaks!! This looks 🔥 great!

jeongsoolee09 committed:
Experiment dealing including / excluding `AddressOfExpr`

jeongsoolee09 committed:
Remove `FALSE_NEGATIVE` labels from offshoots-to-left cases

These are no longer false negatives, is now reported correctly
with the new algorith (`srcSinkLengthMap`).

jeongsoolee09 committed:
Clean up logic in `ArrayDeclaration`

jeongsoolee09 committed:
Fix formatting

jeongsoolee09 committed:
Flag a case as false negative

This is consistently broken across all queries that depend on it:

- c/misra/src/rules/RULE-21-17/StringFunctionPointerArgumentOutOfBounds.ql
- c/misra/src/rules/RULE-21-18/StringLibrarySizeArgumentOutOfBounds.ql
- c/cert/src/rules/ARR30-C/DoNotFormOutOfBoundsPointersOrArraySubscripts.ql
- c/cert/src/rules/ARR38-C/LibraryFunctionArgumentOutOfBounds.ql

jeongsoolee09 committed:
Update the expected results of both queries

This is the aftermath of fixing the formatting.

jeongsoolee09 committed:
Update rule description file of `Memory1`

MichaelRFairhurst commented on cpp/misra/src/rules/RULE-8-7-1/PointerArithmeticFormsAnInvalidPointer.ql:
```diff
@@ -206,7 +220,24 @@ class IndirectUninitializedNode extends Node {
+     *
+     * &
+     *
+     * c.f. `TAddressOfIndex` branch of `PointerFormation`.
```
Nice!

MichaelRFairhurst commented:
LGTM once the build passes!

jeongsoolee09 committed:
Finish the docstring

jeongsoolee09 committed:
Update query metadata and fix formatting

jeongsoolee09 committed:
Super minor formatting

jeongsoolee09 committed:
Merge branch 'main' into jeongsoolee09/MISRA-C++-2023-Memory

jeongsoolee09 committed:
Update unit test results
